[GEP-29] Autoscaling Storage Volumes (#10690)
gardener/gardener: docs/proposals/29-autoscaling-storage-volumes.md

Reviewer: Add this file to the index

Author: Done.

storage management functionality is envisioned in `etcd-druid`. For details, see
[`etcd-druid` feature request issue#481](https://github.com/gardener/etcd-druid/issues/481#issuecomment-1346402013).

### Goals
- General ability to autoscale seed PVs (expand only).
  - Ability is conditional on PVC's storage class supporting resizing.
  - Autoscaling is inactive by default.
  - Ability to selectively activate autoscaling per PVC.
  - Ability to trigger scaling based on relative volume utilisation ratio (`<used bytes> / <capacity bytes>`). Trigger
    threshold configurable via code.
  - Ability to configure absolute upper scaling limit per PVC.
  - Autoscaling causes minimal interruption to scaled workloads.
- Autoscaling ability activated for observability volumes (`prometheus` and `vali`) in seeds'
  `garden` namespace, and in shoot namespaces.
- Absolute upper scaling limit configured on each observability volume.
- If autoscaling is enabled for a newly created volume, reduce the initial size of that volume, as long as such reduction
  does not contradict Gardener's overall reliability goals. For instance, Prometheus instances display sharp storage

>   allocation spikes; the choice of initial size must should accommodate those in an acceptable manner.

Reviewer 2: "must should" is probably not what you intended to write. Please choose either.

Author: Fixed, thanks!

| Shoot          | prometheus           | 20Gi     | 1Gi          | 40Gi     | 600Mi            |
| Shoot          | vali                 | 30Gi     | 1Gi          | 60Gi     | 200Mi            |
| Seed           | prometheus-aggregate | 20Gi     | 2Gi          | 40Gi     | 600Mi            |
| Seed           | prometheus-cache     | 10Gi     | 2Gi          | 20Gi     | 1Gi <sup>*</sup> |
| Seed           | prometheus-seed      | 100Gi    | 5Gi          | 200Gi    | 600Mi            |
| Seed           | vali                 | 100Gi    | 5Gi          | 200Gi    | 600Mi            |

<sup>*</sup> The greater minimal threshold mitigates [risk due to circular dependency](#reliance-on-prometheus). 
### Transition Strategy
The following steps will be executed over time, to minimise the risk of disruption to existing Gardener instances:

1. The proposed scaling approach will be deployed behind a feature gate and disabled by default.
2. The new feature will be gradually enabled on seed-by-seed basis:
   1. The feature gate will be enabled for the seed
   2. Correct operation will be verified via direct observation

> 3. The feature flag will be promoted and eventually removed.

Reviewer 2: We should pre-flag this as recommended approach for Gardener operators. It is not possible to enforce seed-by-seed rollout. Some operators might do it, but others might not even turn the feature gate on and will only get it when it is turned on by default.
Therefore, while this approach is laudable it is just one recommended option and nowhere enforced. The text should make this clear.

Author: I changed the wording from "will" to "have the option to".

---
title: "Autoscaling Storage Volumes - Phase 1: Observability Volumes"
gep-number: 29
creation-date: 2024-10-21
status: implementable
authors:
  - "@andrerun"
reviewers:

>   - "TBD"

Reviewer: You can add me

Author: Done. Thanks!

periodically examines, and if capacity is found to be near exhaustion, `pvc-autoscaler` takes action by updating the
PVC's storage request.

In this initial iteration, `pvc-autoscaler` scales two categories of `vali`, `prometheus`, and `alertmanager` volumes:
those in shoot namespaces, and those in the seed's `garden` namespace.

`pvc-autoscaler` publishes Prometheus metrics to anonymous scrapers via a `/metrics` HTTP endpoint. A `ServiceMonitor`
object is created in the `garden` namespace, and drives the seed's `prometheus-operator` to configure a scrape on that
endpoint by `prometheus-seed`.

`pvc-autoscaler` runs in active/passive replication mode, based on the standard leader election mechanism, supplied by
the K8s controller runtime library. Only the active replica is shown in _Fig.1_.

### PVC Autoscaler - Deployer Structure
The `pvc-autoscaler` application and its supporting artefacts are deployed by `gardener`, as part of the seed
reconciliation flow.

The `vali` and `prometheus` deployers, part of `gardenlet`'s shoot reconciliation flow, ensure that the PVC template section in

> the respective StatefulSet objects contains the annotations, necessary to instruct `pvc-autuscaler` how to act on the

Reviewer: typo

Author: Fixed. Took me surprisingly long to spot.

on existing shoots, and on the observability PVCs of existing seed clusters. To enable PVC autoscaling for existing
shoots, those PVCs will be directly annotated by the operator. Such direct intervention allows gradual transition,
where autoscaling can initially be enabled only for a few PVCs for validation purposes.
Similarly, to enable PVC autoscaling on existing seed clusters' `garden` PVCs, those will be directly annotated at
operators' discretion.

![02-operator_structure.png](assets/gep29-02-operator_structure.png)

_Fig.2: Operator structure_

VPA is used to scale `pvc-autoscaler`. For the future security enhancement scenario, where `kube-rbac-proxy` is used
(see [section 'Future Enhancements'](#future-enhancements)),
that container is expected to be excluded from autoscaling, and to run with fixed resource requests. Pod evictions
driven by that lean container are unjustified, as they would only achieve minimal resource savings, and disrupt
the main container.

The feature is deployed behind a dedicated feature gate. When disabled, `pvc-autoscaler` is removed, and related

> annotations are removed from the shoot control plane observability StatefulSets, and the seed observability StatefulSets.

Reviewer: If I understand correctly, human operators have to annotate PVCs (should better be labels for easier filtering) they want to be auto-scaled. Is this understanding correct? If so, do we need a feature gate at all? Wouldn't removing the label be good enough?

Author: As soon as we enable the new feature, each new shoot, created after that point on, comes out in the new shape (observability volumes automatically annotated, small initial size, and PVC autoscaling enabled). Preexisting shoots are not affected, and require annotating manually, but the shoot creation process is immediately affected.

Reviewer: Interesting. Why are pre-existing shoots not affected? Who/what mechanism will label the volumes? What's the different with new shoots?

Author: >what mechanism

We're creating PVCs via a PVC template in the STS. Not direct reconciliation. So once a PVC is created, the template doesn't affect it. The implementation would update the PVC template with the autoscaler-enabling annotations, and leave the rest to the operator.

I like how that works out in the end, leaving control in the hands of the operator. For one thing, existing PVs are likely heavily underutilised, so enabling autoscaling won't be doing much for them. If the operator wants to save on storage, it's up to them to delete the volumes, so they can start anew, from a small size.

There's actually a piece of existing code in `gardenlet`, which resizes PVCs upon reconciliation, if a new size is specified (as a parameter(value) to the deployer). But I plan to disable it when autoscaling is active. For downsizing, the existing code deletes the PV. For upsizing, it relies on the storage class supporting expansion, which it does not check. So it was just a quick fix.

Reviewer: I don't like the approach that a human operator needs to do things manually beyond activating the feature gate. We always strive for automation where possible and reasonable. In this light, what you propose is inconsistent with what happens for newly created shoots and creates additional overhead. As an operator using this feature gate, I would expect all shoots (new and existing) to be considered for PVA. Hence, `gardenlet` should also update existing PVCs with the label, even if it might not have an effect since most of them might be underutilised.

Author: During offline discussion, @rfranzke expressed preference that `pvc-autoscaler` be refactored to use a new custom object as source of per-PVC settings, instead of the current annotations in the PVC object itself. The goal would be, improve UX by building a parallel with the existing VPA.

In a discussion with the `pvc-autoscaler` authors, they expressed an open stance toward the idea of a per-PVC configuration object as a possible future enhancement, with the note that `pvc-autoscaler` is not under active development at the moment. Delaying the present GEP to wait for a substantial rework of `pvc-autoscaler` thus seems counterproductive: the `gardenlet` feature proposed by the GEP has no user exposure, and can be refactored at any time in the future, at minimal cost.

As a side benefit from the meeting with the `pvc-autoscaler` authors, it was noted that this GEP should explicitly communicate our vision regarding migration of existing storage volumes to autoscaling. To that end, I added a small new section in the GEP, dedicated to that topic. Currently if follows the GEP's original view of deferring all responsibility to the operator, as it seems to coincide with `pvc-authoscaler`' authors' vision. However, this decision should ultimately be made in the present forum. The new section is thus only meant to serve as a basis for discussion, and is open to change.

Reviewer 2: I am very surprised concerning this discussion. From my point of view, this puts too much burden on the operators, especially for large landscapes. We need automation from the beginning.

The proposal has two benefits from my point of view:
1. It potentially allows for cost reduction due to restricted volume usage based on real usage
2. It allows longer retention of observability data in highly loaded scenarios

As long as we do not consider scaling down volumes, 1. only has an effect for short lived clusters.

The second benefit however could also have a positive effect for existing clusters, but only if we roll it out properly. Manual operator action is no real option for large landscapes in this context. Automation is required and should be included in the GEP.

Reviewer 3: I also believe that autoscaling should be enabled for all PVCs targeted by this GEP.
Including what was stated already in that direction,  this would allow us to check whether `pvc-autoscaler` can handle the increased load on large landscapes right away.

P.S. If we leave it to operators, they might introduce inconsistencies, enable the annotation on volumes we do not expect it to be enabled on and etc... It's better to have the enablement tested and as part of our code base.

Author: @rfranzke @ScheererJ @plkokanov 
A consultation with with observability folk (@nickytd, @dnaeon) confirmed that there are currently no plans to provide automatic migration for preexisting PVs.  Obviously we can't just drop the data either. Therefore the best we can do at this time is what you requested above: automatically enable autoscaling on preexisting PVs (and we leave them at their current size).

In the latest commit I updated the GEP to automatically annotate existing `vali` and `prometheus` PVCs.

Reviewer: It is great that handling existing PVCs are in scope now. Yet, I am unhappy with the refusal of the discussed enhancement of PVA w.r.t. switching its annotation APIs to CRDs. In an offline discussion, @ScheererJ shared similar thoughts, but perhaps he can also speak up here.

> In a discussion with the `pvc-autoscaler` authors, they expressed an open stance toward the idea of a per-PVC configuration object as a possible future enhancement, with the note that `pvc-autoscaler` is not under active development at the moment. Delaying the present GEP to wait for a substantial rework of `pvc-autoscaler` thus seems counterproductive: the `gardenlet` feature proposed by the GEP has no user exposure, and can be refactored at any time in the future, at minimal cost.

I disagree with the statement that switching from annotations to CRD is "substantial rework" or "counter-productive". It is also no excuse that this component is "not under active development". Such a change can be done with a small number of development hours. This is still our component in the `gardener` org, so we have full control of it and can decide to put it into "active development" for a day or two if you will.

As you seem to agree that this is actually a good thing/a reasonable enhancement, I don't see any satisfying reason to not to this now before PVA gets integrated and rolled out to existing systems. Refactoring this later, even at "minimal cost", is just waste and extra-effort. It can be done already now, effectively saving additional work in the future.

Author: >As you seem to agree that this is actually a good thing(...)

The difficulty when joining data from multiple sources is something which caused me to reverse my preference for this case.

In general, I favour modularisation, and data models and REST APIs are no exception. In our case, that modular design would be the one based on a CRD. However, I recently realised that the lack of support for joining data makes 1:1 object relationships quite harmful to UX in the K8s CLI case (kubectl). Splitting objects into custom resources turns one-liners into highly inconvenient for+for+if manual joins. 

It's also worth noting that the analogy with VPA isn't complete, from a UX perspective. VPA is not a small type, and Pod is particularly large. So merging those would make for an overwhelmingly large object. PVC on the other hand is a very small object, and a PVCAutoscaler spec would consist of (IIRC) just 5 primitive values, 4 of which are very optional. Aside from reporting back status, PVCAutoscaler is likely to end up as an entire K8s resource which configures just "Enabled = true". 

A dedicated CRD would allow for a much better extensibility. IMO, a CRD would be particularly useful if we decide to implement multiple scaling strategies/algorithms, each with a different set of parameters. However, I do not know that we would make use of such extensibility.

So, all things considered, if I was faced with the choice whether to base `pvc-autoscaler` on annotations or a CRD, I'd be on the fence. Since we already have implementations based on one of those options, I find it more economical to go with that option. In other words, follow the original plan to deliver a MVP, and only rework/extend if we find it necessary.

Reviewer 2: @andrerun Sorry, but I do not get your points.

Annotations are a non-typed interface, which can easily pose problems. Just think of a little typo in the key or value, which is very easy to overlook. Even if there are currently only 5 primitive values to configure that certainly makes sense to be put into a separate CRD so that you can configure it properly with type safety and checks.

How would you even envision this with annotation? Would you put five annotations on a resource or have a `json` value for an annotation? Both approaches are rather bad from my point of view.

However, just do not only take my word for it. Feel free to check what Tim Hockin for example says with regards to annotations in the ingress API in the [kubernetes podcast](https://kubernetespodcast.com/episode/041-ingress/).

Especially as you have put it as the long term vision to also provide volume autoscaling for shoot workloads, I think it is necessary to have a proper API (read: CRD) and not just annotations. It also directly allows you to provide a proper status to your consumers.

From my point of view, the current approach with annotations is a blocker. Please reconsider.

Reviewer 4: I am sharing my thoughts on the CRD vs annotations discussion without imposing a blocker for a rework of the existing pvc-autoscaler:

A major pro of the CRD approach is its status. I haven't read the full GEP, neither looked into the implementation of the pvc-autoscaler. But, my assumption is that the pvc-autoscaler creates events on successful/failed resize operations. If that's the case, events are ephemeral and they get deleted when they ttl expires (defaults to 1h for a Gardener Shoot). When a failed resize operation occurs, then the pvc-autoscaler can reflect this in the CR status. Then, other components like gardenlet can health check the custom resource. This is similar to how gardenlet is health checking VPA custom resources and reflecting issues with them in the Shoot status. Otherwise, how can an operator can be noticed about failed resize operations that occur in the landscape?

tl;dr: IMO, a bigger pro argument for the CRD approach is not its spec, but rather it is status. An operator can read the status for investigation/debugging purposes. gardenlet can read the status and then reflect failures to the Shoot status.

Reviewer 5: > @andrerun Sorry, but I do not get your points.
> 
> Annotations are a non-typed interface, which can easily pose problems. Just think of a little typo in the key or value, which is very easy to overlook. Even if there are currently only 5 primitive values to configure that certainly makes sense to be put into a separate CRD so that you can configure it properly with type safety and checks.
> 
> How would you even envision this with annotation? Would you put five annotations on a resource or have a `json` value for an annotation? Both approaches are rather bad from my point of view.
> 
> However, just do not only take my word for it. Feel free to check what Tim Hockin for example says with regards to annotations in the ingress API in the [kubernetes podcast](https://kubernetespodcast.com/episode/041-ingress/).
> 
> Especially as you have put it as the long term vision to also provide volume autoscaling for shoot workloads, I think it is necessary to have a proper API (read: CRD) and not just annotations. It also directly allows you to provide a proper status to your consumers.
> 
> From my point of view, the current approach with annotations is a blocker. Please reconsider.

Hey @ScheererJ ,

I don't see how this is any different from the existing implementation of Network Policies controller, which is managed by a mix of labels + annotations, and on top of that -- embeds JSON within these annotations.

References:

- https://gardener.cloud/docs/gardener/network_policies/
- https://github.com/gardener/gardener/issues/7352

This is an example annotation which magically configures a network policy.

``` yaml
annotations:
  networking.resources.gardener.cloud/from-all-garden-scrape-targets-allowed-ports: '[{"port":<metrics-port-on-pod>,"protocol":"<protocol, typically TCP>"}]'
```

According to what you've mentioned this is a bad design, however we are already promoting such design (and implementation) to Gardener users with the use of the network policies annotations + labels.

I'm not picking on the Network Policies controller here, I'm simply trying to understand the rationale here and the criteria against which you are measuring one annotations-approach as good, and another one as bad? @ScheererJ , @rfranzke ?

For the sake of moving forwarding with the integration of pvc-autoscaler into g/g I've also merged CRDs support into `pvc-autoscaler`.

However, please elaborate on the question I've asked above, so that myself (and others included) can better understand the design decisions and rationale which make annotations sometimes good, and other times bad.

Thanks!

Reviewer 2: Hey @dnaeon,

your example is spot on. Some of the network policy controller annotations use structured data (read: JSON) and I am not too happy about it. Unfortunately, it is not always clear cut in good/bad. As usual, it depends.

Why do I find it more acceptable (but still not great) to do it like this for the network policy controller? This API is purely for Gardener internal usage. Yes, extension implementors may need to deal with it, but that is rather a limit amount of people. In addition to that, there is no need to provide any status values.

The network policy controller is meant for simplifying the network policy creation as it is hard to create the correct network policies (as history/experience shows). The abstraction over the network policy API was done with annotations/labels and evolved from simple to more complicated use cases. If I remember correctly, the initial solution did not include the complex JSON values, but @rfranzke may remember better.

If we would consider making the network policy controller available to cluster owners I would probably lean more towards a new CRD instead of exposing our label/annotation API.

TLDR: It always depends. The bigger your potential API audience is the simpler/cleaner/better your API should be. PVC autoscaling has large goals (all Gardener clusters) while network policy controller only wants to rule the Garden/Seeds.

Hope that helps even though it is not clear cut.

Thanks a lot for implementing the CRD support in `pvc-autoscaler`. Let's adapt the GEP accordingly and finish the remaining smaller discussions so that we can move forward.

@andrerun Please check the commit messages. The bot is currently complaining, which would block a merge.

Author: To record an answer which was part of an offline discussion:
`pvc-autoscaler` uses a status condition on the PVC to reflect the error/success status of the last resize.

Author: Gardener core team effectively resolved this discussion by posting an overruling requirement to use the approach based on a custom resource. For details, see [this comment](https://github.com/gardener/gardener/pull/10690#issuecomment-2535767407).

#### Initial volume size and MaxAllowed

| Component type | Component            | Old size | Initial size    | Max size |
|----------------|----------------------|----------|-----------------|----------|
| Shoot          | prometheus           | 20Gi     | 1Gi             | 40Gi     |
| Shoot          | vali                 | 30Gi     | 1Gi             | 60Gi     |
| Seed           | prometheus-aggregate | 20Gi     | 2Gi             | 40Gi     |
| Seed           | prometheus-cache     | 10Gi     | 6Gi<sup>*</sup> | 20Gi     |
| Seed           | prometheus-seed      | 100Gi    | 5Gi             | 200Gi    |

> | Seed           | vali                 | 100Gi    | 5Gi             | 200Gi    |

Reviewer: Where are the max sizes coming from?

Author: It's just 2x current size. I used a very sound scientific approach: I more or less closed my eyes and tossed a coin. I'm absolutely open to better suggestions. The factors that affected my choice are:
- I want to allow the autoscaler to provision more than current storage, where needed.
- With an autoscaler that's not combat-proven, I'm concerned that it might have a bug which causes it to unnecessarily expand volumes all the way to the max size. And there's no support for shrinking. So I'm worried about the scenario where we're forced to delete all observability volumes, just so we can restore them to acceptable size.

So I allowed the autoscaler the freedom up to 2x current size. My thinking is: 2x should suffice for a couple of months. And after that time, we'll have gained confidence in the reliability of `pvc-autoscaler` and will be comfortable increasing the max value to its ultimate level.

   1. The feature gate will be enabled for the seed
   2. The correct operation of newly created shoots will be verified via direct observation
   3. The observability PVCs of preexisting shoots will gradually be annotated for scaling by `pvc-autoscaler`
   4. The correct operation of preexisting shoots will be verified via direct observation
3. The feature flag will be promoted and eventually removed.

The following metrics will be recorded before and/or after the transition of the seed to the new functionality,
in order to evaluate the economical effect, and the reliability of the scaling solution:
   - Average utilisation of observability volumes:
     - Preexisting volumes (before and after)
     - Newly created volumes (after)
   - Percentage of observability volumes above 90% utilisation (before and after)
   - Utilisation of observability volumes at p10, p50, p90:
     - Preexisting volumes (before and after)
     - Newly created volumes (after)
   - Peak volume utilisation (after)
   - Observability volume size: at p10, p50, p90, and average size (after)
   - Observability volume average size and count (before)

> ### Future Enhancements

Reviewer: What about plans to scale ETCD and shoot volumes? The headline says "Phase 1", so there are other phases and more plans?

Author: TL;DR: Next phase = PVC autoscaling capability for customer (shoot) volumes.

Providing autoscaling capability to shoot PVs absolutely is in the vision. We don't have a fixed execution plan yet, but we're quite certain that we want it. The high level "plan" is, deploy for internal observability volumes (current phase), use that to gain confidence/maturity, then offer the functionality to customer (shoot) PVs.

I haven't considered ETCD, so I don't know yet if it would be a good candidate. I know that it has an 8GB limit to the data it can store, but I guess the WAL file could grow substantially larger, under heavy load and delayed compactions. On the other hand, it may be wasteful to inflate the volume due to a one-time high-load event inflating the WAL, and then pay for a large volume ever after.

Reviewer: @andrerun: Please add these thoughts to the document :)
@unmarshall @shreyas-s-rao and colleagues: Thoughts about scaling the ETCD volume?

Reviewer 6: You can have a look at:
https://github.com/gardener/etcd-druid/issues/481#issuecomment-1346402013
The prerequisite of doing this for ETCDs is https://github.com/gardener/etcd-druid/issues/588 (we are in the process of writing the DEP for it). For etcd clusters it is therefore recommended that you make changes via the `Etcd` custom resource and delegate the responsibility to etcd-druid to resize the PVs for etcd cluster. etcd-druid will ensure that the quorum is maintained at all times.

Author: Done. Notes regarding future plans (scaling customer volumes, and not scaling Etcd volumes), are now available in the paragraph starting with "The current initiative is limited to scaling observability components", and the paragraph right after it.

2. `metrics-server` polls `kubelet` resource metrics. Default poll interval: 1 minute
   ([source](https://github.com/kubernetes-sigs/metrics-server/blob/master/FAQ.md))
3. `kube-apiserver` forwards metrics API calls to `metrics-server`, which is the registered APIService for metrics. Call
   forwarding is instantaneous, so the delay is minimal.
4. The seed's cache Prometheus polls the metrics provided by `kube-apiserver`. Default poll interval: 1 minute.
5. `pvc-autoscaler` polls the seed's cache Prometheus instance. Poll interval: 1 minute.

The reaction time, until `pvc-autoscaler` **initiates** a volume resize, can thus exceed 4 minutes. The resize operation
itself, including the file system resize, can further take a few minutes. The cumulative delay, roughly estimated to up
to 10 minutes, is mitigated via an appropriate safety margin, provided by the choice of utilisation level, configured
as a scaling trigger threshold.

## References
- [[1]] `pvc-autoscaler` GitHub repo
- [[2]] `pvc-autoscaler` evaluation of existing autoscalers
- [[3]] `kube-rbac-proxy` GitHub project
- [[4]] Kubernetes TokenReview API
- [[5]] Kubernetes SubjectAccessReview API

> [1]: https://github.com/gardener/pvc-autoscaler

Reviewer 7: `pvc-autoscaler` repo is still private. Is it required for the purpose of this review?

Author: Formally - no, but with the repo private, the GEP would be suggesting the use of some obscure, completely undocumented autoscaler.

So if we feel that the `pvc-autoscaler` repo is already good enough to publish, that may be the easiest path. Next option is, publish just the binary and documentation. Last option (still viable): publish just the binary and live with consuming an obscure application for a while.

Author: [pvc-autoscaler](https://github.com/gardener/pvc-autoscaler) repo is now public. :tada: Thanks, @dnaeon, @JordanJordanov , @nickytd  for the instantaneous reaction!

> [2]: https://github.tools.sap/I746790/notes/tree/master/pvc-autoscaler#evaluate-existing-pvc-autoscalers

Reviewer 7: This is an internal resource, please, remove it.

Author: Done. Thanks for catching this!

  * [Non-Goals](#non-goals)
* [Proposal](#proposal)
  * [PVC Autoscaler - Runtime Structure](#pvc-autoscaler---runtime-structure)
  * [PVC Autoscaler - Deployer Structure](#pvc-autoscaler---deployer-structure)
    * [Initial volume size and MaxAllowed](#initial-volume-size-and-maxallowed)
  * [Transition Strategy](#transition-strategy)
  * [Future Enhancements](#future-enhancements)
* [Discussion and Limitations](#discussion-and-limitations)
  * [Alternatives](#alternatives)
  * [Limitations](#limitations)
* [References](#references)
<!-- TOC -->

## Summary
A "one size fits all" strategy is a poor fit for the persistence volumes of Gardener observability workloads.
This document proposes a simple design to enable flexible, dynamic PV resizing in cases where the underlying
infrastructure supports it.
 
## Motivation
Logs are vital for understanding and troubleshooting Gardener clusters. Yet we operate under tight resource constraints
and logging stack PVs default to a static size of 30GB, for ALL shoots. To prevent reaching this limit and bringing down
the logging stack, there is a curator sidecar which periodically removes older logs. As a result, in some very large or

> busy clusters, this strategy allows retention of logs for only for 2-3 days.

Reviewer 3: ```suggestion
busy clusters, this strategy allows retention of logs only for 2-3 days.
```

Author: Fixed, thanks!

On the other hand, while some clusters experience storage starvation, others operate with just 5-10% of their observability
volumes utilised.

The present initiative is focused on observability storage volumes, and aims to:
- Optimise volume utilisation
- Ensure that high-throughput clusters receive sufficient observability storage space

The goals are achieved by the following means:
- A Gardener controller is introduced to auto-scale PVs (expand only)
- Observability volumes are selectively marked for action by said controller
- Initial volume size is reduced

> - Where the storage infrastructure does not support volume resizing, the preexisting, static-size volume sizing strategy is employed

Reviewer 3: `gardenlet` is what sets the static-size volume size, how will it check whether the underlying infrastructure supports volume expansion or not?

Author: The capability is expressed by the storage class:
```
apiVersion: storage.k8s.io/v1
kind: StorageClass
allowVolumeExpansion: true
(...)
```
So my initial idea is to check whether the default storage class supports expansion. This allows for a problem f the default storage class changes from one which supports expansion to one which does not, but that seems very unlikely to happen outside development environments.

The current initiative is limited to scaling observability components - _Vali_ and _Prometheus_.
However, the long term vision is, once the PV autoscaling implementation matures enough, to also use it to offer PV
auto-scaling for shoot (i.e. customer) volumes.

There are no plans to extend the present initiative to shoot Etcd volumes. Instead, a more advanced, Etcd-specific
storage management functionality is envisioned in `etcd-druid`. For details, see
[`etcd-druid` feature request issue#481](https://github.com/gardener/etcd-druid/issues/481#issuecomment-1346402013).

### Goals
- General ability to autoscale seed PVs.
  - Ability is conditional on PVC's storage class supporting resizing.
  - Autoscaling is inactive by default.
  - Ability to selectively activate autoscaling per PVC.
  - Ability to trigger scaling based on relative volume utilisation ratio (`<used bytes> / <capacity bytes>`). Trigger

>     threshold configurable via code.

Reviewer 2: I sure hope we do not want to hardcode the thresholds.

```suggestion
    threshold configurable via configuration.
```

Author: Per offline discussion, configuration values will be in gardenlet code, similar to workloads' resource request values.

  - Ability to configure absolute upper scaling limit per PVC.
  - Autoscaling causes minimal interruption to scaled workloads.
- Autoscaling ability activated for observability volumes (`prometheus` and `vali`) in seeds'
  `garden` namespace, and in shoot namespaces.

> - Absolute upper scaling limit configured on each observability volume.

Reviewer 2: Why is the upper scaling limit mentioned twice? What is the difference?

Reviewer 3: I think the first mention means "the ability to configure it" and here it is meant that "it is configured by `gardenlet` when deploying observability PVCs"

Author: @plkokanov, spot on! The former goal is about a general capability to constrain expansion.

The "General ability to autoscale seed PVs" section embodies the vision that PVC autoscaling will go beyond observability volumes in the future. Then follows the group of goals, specific to the current phase, tied to observability volumes.

Author: To highlight the few discussion threads which remain open, I'm closing threads which appear to have been addressed. If you find that anything needs more discussion/work, please reopen.

> - Reduce the initial size of a volume, when autoscaling is activated for that volume, if such reduction is aligned with

Reviewer 3: Maybe this should be rephrased to indicate that the current goal is for reduction of initial size to only happen for new volumes?

Author: Done.

>   general Gardener goals (cost, reliability).

Reviewer 2: What does this mean? When is a reduction not aligned with the "general Gardener goals"?

Author: A reduction of initial requests is always aligned with the cost goal, but can hurt reliability. Case in point, the seed `cache-prometheus` has spikes where storage usage grows with ~600MB over 15 seconds. No autoscaler can react that fast, so I'm mitigating the issue by setting a larger initial request for that PVC.

Reviewer 2: So let's make the GEP more understandable by adapting this sentence to include the additional info you provided. What you stated in your answer is really hard to make out from the sentence in the GEP.

Author: I changed the goal to the following wording:
> If autoscaling is enabled for a newly created volume, reduce the initial size of that volume, as long as such reduction does not contradict Gardener's overall reliability goals. For instance, Prometheus instances display sharp storage allocation spikes; the choice of initial size must should accommodate those in an acceptable manner.

Will be available with next push.

Author: To highlight the few discussion threads which remain open, I'm closing threads which appear to have been addressed. If you find that anything needs more discussion/work, please reopen.

> - Some metrics, which represent an overview of the PVC autoscaling state per seed, available through one or more of the seed

Reviewer 3: ```suggestion
- Metrics, which represent an overview of the PVC autoscaling state per seed, available through one or more of the seed
```

Author: I'm intentionally being explicit with the quantifier (some), to avoid miscommunication. I'm currently not making a promise that all metrics will be made available. The initial prototyping work I did publishes all metrics, but I'd like the option for the implementer to drop some of the metrics, if they judge them unnecessary.

Reviewer 2: "Metrics" also does not include "all" and is more readable. I think you still have enough options while retaining readability. Please adapt.

Author: Changed to "A set of metrics"

  system's Prometheus instances.
- Provide operational documentation for Gardener Operators.

> - Ability to enable/disable the feature per seed, via feature gate.

Reviewer 2: Should existing volumes be autoscaled if the feature gate is enabled, i.e. will existing volumes grow beyond their previous size with this?

Reviewer 3: Good point @ScheererJ. Could this also mean that if a volume has grown beyond their previous size (even for new clusters) and the feature gate is disabled, the volume will be recreated as part of https://github.com/gardener/gardener/blob/5417a7bbb3adab9d192abd8e8e2ac4b7fe7c9a2d/pkg/component/observability/logging/vali/vali.go#L152-L156

Reviewer 2: Also, what happens if we create a cluster while autoscaling is enabled, i.e. with a small pvc size, and autoscaling is disabled at a later point in time. Will the pvc increased to the "default" or will it stay "small"?

Author: @ScheererJ Yes. After the recent update to the GEP, autoscaling is enabled on preexisting volumes

@plkokanov Yes, after the following sequence of events:
1. Autoscaling feature enabled
2. Autoscaling enabled on volume X
3. Autoscaler expands volume X beyond the static level of capacity, which would be allocated if the feature was  disabled
4. Autoscaling feature disabled
...the volume would be deleted and redeployed with capacity at the static level.

@ScheererJ It will be expanded to "default" capacity.
If the autoscaling feature is disabled, a volume with a small size will be resized by the code @plkokanov mentions above. If the storage class still supports resize at that time, it will be expanded to the "default" static capacity.

### Non-Goals
- Ability to activate autoscaling on a group of PVCs, other than specifying each PVC individually. Activate by selector
  or namespace is not supported.
- Ability to selectively deactivate autoscaling, e.g. by PVC selector or namespace.
- Ability to scale a PVC if the storage class does not support resizing.
- Ability to deactivate autoscaling for any of the observability volumes specified in the 'Goals' section.

> - Ability to scale volumes in shoot clusters.

Reviewer 2: Not sure why deactivation of autoscaling is mentioned twice, but not once any mention of downscaling. I thought that this was an explicit non-goal.

WDYT?

Author: >deactivation (...) mentioned twice

The latter non-goal is indeed formally sufficient. But I strive to differentiate the case of group deactivation from that of individual object overrides. Both are meaningful features - tempting to have, so I want to be clear that we won't have them in the current phase.

>not once any mention of downscaling

I was on the fence about whether to describe that as a non-goal, or an implementation limitation. Your vote breaks the tie, so I'll add to non-goals.

Author: To highlight the few discussion threads which remain open, I'm closing threads which appear to have been addressed. If you find that anything needs more discussion/work, please reopen.

### Notes:
- Shrinking a PVC is not supported. A limitation imposed by the underlying infrastructure, PVC size can only be

>   increased, and not decreased.

Reviewer 2: Is it not possible to work with snapshots and creation of a new volume to achieve decreasing size of volumes?

Whether we want to do it is another question, but I suppose it is neither rocket science nor unheard of.

Reviewer 5: [Shrinking persistent volumes is not supported](https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/) by Kubernetes.

In theory, yes we can implement shrinking on our own by leveraging snapshots, migrating data, etc. But this in turn creates additional complexity IMO. It also creates potential for possible downtimes during data migration to a smaller pv.

While this is something we can further discuss (and potentially implement) I think it should not be part of this GEP.

Reviewer 2: Like I said, I have no problem if we do not want to address it initially or at all, but I would expect it to be listed as non-goal (if we do not want to do it initially) and as a future topic (in case we see potential to do it later).

The current note makes it look like we cannot do anything about it, which I believe to be not true. It is rather a decision to take.

Author: >It is rather a decision to take.

Either non-goal or a future topic 

Excellent points! I added it to non-goals.

- The newly introduced auto-scaling mode is predicated on the underlying infrastructure supporting volume resizing.
  This ability is expressed through the K8s StorageClass type. The respective StorageClass must have

>   `allowVolumeExpansion: true`. For details, see the K8s volume expansion documentation.

Reviewer 2: Please add a link.

Reviewer 5: Posting this here for reference.

* https://kubernetes.io/docs/concepts/storage/storage-classes/#allow-volume-expansion
* https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/

Otherwise, I agree - please add these links to the GEP.

Reviewer 2: Yes, please add those links to the GEP.

Author: Done. Will be available with next push.
@dnaeon, thanks for the links!

  
## Proposal
### PVC Autoscaler - Runtime Structure
The overall runtime structure of the component is outlined in _Fig.1_:

> ![01-runtime_structure.png](./assets/gep29-01-runtime_structure.png)

Reviewer 2: If you use abbreviations in diagrams please introduce them beforehand. PVA is not a common abbreviation. It is also used never used in the document except in two diagrams. Maybe using the full name is also an option.

Author: Thanks for spotting! I fixed the two diagrams. Will be available with next push.

_Fig.1: Runtime structure_

For the role of a storage scaling controller, the existing, recently implemented Gardener PVC autoscaler application

> [[1]], hereafter referred to as `pvc-autoscaler`, is used. It is

Reviewer 2: Process question:
Should we not first discuss the general approach via a GEP before diving into the implementation?

Prototyping is obviously required before writing a GEP, but this sentence sounds like we already implemented a production-grade a PVC autoscaler for this purpose, which seems to contradict the GEP process as far as I understand it.

Reviewer 5: The `pvc-autoscaler` is a separate controller, which resides in a separate repo on it's own. The design decisions about it have been discussed already with involved parties.

This GEP discusses the integration of `pvc-autoscaler` into g/g, for which there will be prototyping work to be done.

Reviewer 2: Let me phrase my question differently: Did we have another scenario in mind when designing/implementing `pvc-autoscaler`, which is not related to gardener/gardener?

Frankly, I see it as one consolidated/coordinated story and not two random events happening shortly after one another. However, I may not have the full picture. Feel free to provide more insights.

Reviewer 2: While this has not been clarified, let's hope we manage to make this more clear in the near future within the NeoNephos Foundation.

integrated as a seed system component, and runs as a deployment in the seed's `garden` namespace.
Its primary driving signal is the PVC metrics from the seed's cache Prometheus instance, which it
periodically examines, and if capacity is found to be near exhaustion, `pvc-autoscaler` takes action by updating the
PVC's storage request.

In this initial iteration, `pvc-autoscaler` scales two categories of `vali` and `prometheus` volumes:

> those in shoot namespaces, and those in the seed's `garden` namespace.

Reviewer 3: What about observability components in the `garden` cluster? Have you checked if it would be worth it to add pvc autoscaling there as well?

Author: Yes. I considered it, but I think it's not worth doing at this time. There are few observability volumes in the garden cluster and, at least in productive gardens, the utilisation doesn't vary greatly. So preallocating a bit of extra space works well enough for the garden.

Once the feature is mature, I expect it will be worth adding to the garden cluster. But I expect that it will still have a relatively low priority because of the low volume count there.

Author: To highlight the few discussion threads which remain open, I'm closing threads which appear to have been addressed. If you find that anything needs more discussion/work, please reopen.

`pvc-autoscaler` publishes Prometheus metrics to anonymous scrapers via a `/metrics` HTTP endpoint. A `ServiceMonitor`
object is created in the `garden` namespace, and drives the seed's `prometheus-operator` to configure a scrape on that
endpoint by `prometheus-seed`.

`pvc-autoscaler` runs in active/passive replication mode, based on the standard leader election mechanism, supplied by
the K8s controller runtime library. Only the active replica is shown in _Fig.1_.

### PVC Autoscaler - Deployer Structure
The `pvc-autoscaler` application and its supporting artefacts are deployed by `gardener`, as part of the seed
reconciliation flow.

The `vali` and `prometheus` deployers, part of `gardenlet`'s shoot reconciliation flow, ensure that the PVC template section in
the respective StatefulSet objects contains the annotations, necessary to instruct `pvc-autoscaler` how to act on the
PVCs, created from those templates.

> That `gardenlet` action mechanism, based on PVC templates, only affects PVCs upon creation. Therefore, it has no effect

Reviewer 2: ```suggestion
This `gardenlet` mechanism, based on PVC templates, only affects PVCs upon creation. Therefore, it has no effect
```

Author: Thanks for spotting the error! Correction will be available in the next push.

on existing shoots, and on the observability PVCs of existing seed clusters. To enable PVC autoscaling for existing
shoots, those PVCs will be directly annotated by the operator. Such direct intervention allows gradual transition,
where autoscaling can initially be enabled only for a few PVCs for validation purposes.
Similarly, to enable PVC autoscaling on existing seed clusters' `garden` PVCs, those will be directly annotated at
operators' discretion.
For details regarding the transition of existing volumes to autoscaling, see 
[section 'Enabling Autoscaling for Existing Volumes'](#enabling-autoscaling-for-existing-volumes)

> ![02-operator_structure.png](assets/gep29-02-operator_structure.png)

Reviewer 2: I find it strange to use "Deploy" as the relation between statefulsets and pvcs. An active component like `gardenlet` can deploy kubernetes resources. A statefulset is a kubernetes resource and by itself will not deploy a pvc. Maybe it triggers the creation. Anyway, I would propose to change the diagram to something more fitting.

Author: It depends on the perspective one takes. I've taken the view that a StatefulSet is not just the passive document, but also includes the associated behaviours.

It's not an uncommon view. Even some of the K8s documentation takes it, e.g. https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/:
>A StatefulSet runs a group of Pods

>Manages the deployment and scaling of a set of Pods], and provides guarantees about the ordering and uniqueness of these Pods

Strictly from a pragmatic point of view, in the diagram nothing will be as short and clear as "Deploys".

Author: To highlight the few discussion threads which remain open, I'm closing threads which appear to have been addressed. If you find that anything needs more discussion/work, please reopen.

_Fig.2: Operator structure_

VPA is used to scale `pvc-autoscaler`. For the future security enhancement scenario, where `kube-rbac-proxy` is used
(see [section 'Future Enhancements'](#future-enhancements)),
that container is expected to be excluded from autoscaling, and to run with fixed resource requests. Pod evictions
driven by that lean container are unjustified, as they would only achieve minimal resource savings, and disrupt

> the main container.

Reviewer 2: Not sure if that is not too much detail for a GEP. It seems standard best practice to use VPA and scale only by checking the relevant containers.
WDYT?

Author: Too much detail indeed. The reason this is here, is because this GEP started as a different type of document: a design specification.

All I can do under the circumstances is, delete it. I'd hate to lose design details, but we currently don't have a good place for them, if not in the GEP.

Author: I extracted the secure metrics section into external document. I removed the details from here.

The feature is deployed behind a dedicated feature gate. When disabled, `pvc-autoscaler` is removed, and related
annotations are removed from the shoot control plane observability StatefulSets, and the seed observability StatefulSets.

> #### Initial volume size and MaxAllowed

Reviewer 2: Why is the indentation at level 4? I would have expected this to be a sibling to the surrounding sections.

Author: Because I view it as part of the deployer behaviour. The specific values are impactful enough to be worth listing, but I don't want to promote the table to a higher level. Impactful as it is, it's still a fairly low level implementation detail.

However, your comment made me reconsider the title for the parent section. Instead of "PVC Autoscaler - Deployer Structure", I think I should use just "PVC Autoscaler - Deployer", since the content of the section actually goes beyond static structure, and into behaviour/runtime effects. WDYT?

| Component type | Component            | Old size | Initial size    | Max size |
|----------------|----------------------|----------|-----------------|----------|
| Shoot          | prometheus           | 20Gi     | 1Gi             | 40Gi     |
| Shoot          | vali                 | 30Gi     | 1Gi             | 60Gi     |
| Seed           | prometheus-aggregate | 20Gi     | 2Gi             | 40Gi     |
| Seed           | prometheus-cache     | 10Gi     | 6Gi<sup>*</sup> | 20Gi     |
| Seed           | prometheus-seed      | 100Gi    | 5Gi             | 200Gi    |

> | Seed           | vali                 | 100Gi    | 5Gi             | 200Gi    |

Reviewer 2: Did you consider the impact on performance due to the smaller size? Depending on the infrastructure, the amount of guaranteed i/o operations may depend on the volume size. Hence, smaller volumes may lead to lower performance. Is this sufficient? If not, how should that be handled?

Author: Excellent point! I had based the proposal on the overall nature of the observability workloads, which to my knowledge use log-structured DBs, and thus should be expected to do mostly sequential I/O. Plus a measurement of the data volume load patterns.

Your question prompted me to do a more detailed analysis of real-life instances. It's not a trivial matter, but from what I see, we should be ok with the smaller volumes. Prometheus does a small number of relatively larger writes. Overall it is not too storage-intensive. Vali is the demanding case. It has noteworthy sustained load, and some very sharp spikes. With it, per-GB quota on GCP would not alone suffice to cover the sustained load, and the pod would rely on some of the node's baseline IOPS quota. But even with a node that's loaded only with Vali pods, the baseline quota should suffice, if my guesstimates are correct.

In the extreme case where a shoot is created and immediately loaded to generate extreme amount of logs, it could use a substantial fraction of the node's baseline quota (realistically, about 1/3) until its volumes fill up and trigger expansion (which would come up with its own per-GB quota). Even then I don't expect a practical problem, except in case of deliberate DoS attack, and even that I expect would result in just a bit of delay in log collection.

As to the extreme load spikes I see, those will definitely dip into node/burst quotas, but that's mostly what those quotas are for, anyway.

<sup>*</sup> The greater initial size mitigates [risk due to circular dependency](#reliance-on-prometheus). 
### Transition Strategy
The following steps will be executed over time, to minimise the risk of disruption to existing Gardener instances:

1. The proposed scaling approach will be deployed behind a feature gate and disabled by default.
2. The new feature will be gradually enabled on seed-by-seed basis:
   1. The feature gate will be enabled for the seed
   2. The correct operation of newly created shoots will be verified via direct observation
   3. The observability PVCs of preexisting shoots will gradually be annotated for scaling by `pvc-autoscaler`
   4. The correct operation of preexisting shoots will be verified via direct observation

> 3. The feature flag will be promoted and eventually removed.

Reviewer 2: This implicitly mixes too many personas (developer/operator). Therefore, it is not very helpful.

To make it concrete: Will you check that newly created shoot clusters in any Gardener landscape are working? Most likely not. Why do you expect operators to do this?

As mentioned in above discussion, manual action on a large scale are no option. Please propose a solution that is in line with this.

Author: >Will you check that newly created shoot clusters in any Gardener landscape are working?

Yes, that is what I propose as part of the validation workflow, in addition to standard validation.

Now that we automatically enable autoscaling on preexisting volumes, we no longer have the ability to gradually affect preexisting shoots. The ability to roll out on a seed-by-seed basis is still valuable in reducing risk.

   - Average utilisation of observability volumes:
     - Preexisting volumes (before and after)
     - Newly created volumes (after)
   - Percentage of observability volumes above 90% utilisation (before and after)
   - Utilisation of observability volumes at p10, p50, p90:
     - Preexisting volumes (before and after)
     - Newly created volumes (after)
   - Peak volume utilisation (after)
   - Observability volume size: at p10, p50, p90, and average size (after)
   - Observability volume average size and count (before)

#### Enabling Autoscaling for Existing Volumes
Autoscaling is not automatically enabled for storage volumes which already exist at the time when the new feature is
enabled. That applies to both volumes in shoot namespaces, and those in the `garden` namespace.

These volumes have large, fixed capacity, which in most cases is severely underutilised. Merely enabling autoscaling
on them would do nothing to release the unused storage space. It is the operator's responsibility to
decide which of those volumes to migrate to autoscaling, whether to release the unused space, and if so, whether

> to preserve the existing data on the volumes, and how.

Reviewer 2: As mentioned above, I think that this is not a good approach and should be reconsidered.

Putting this responsibility on the operator is a clear no-go from my point of view.

Author: This is resolved in the latest version of the GEP. I added a note that autoscaling will be automatically enabled on preexisting volumes.

### Future Enhancements
#### Metrics authorization
`pvc-autoscaler` supports access control over its `/metrics` endpoint. This feature will not be utilised by the first
implementation round. It is envisioned as a future enhancement. In that operation mode,
the primary `pvc-autoscaler` container publishes unauthorized, plain text metrics only on the loopback interface.
Peer pods are configured to look for `/metrics` at a different port, which is served by a secondary container, running
the `kube-rbac-proxy`[[2]] application. It serves HTTPS and authenticates and authorizes each request against the runtime
cluster via the TokenReview[[3]] and SubjectAccessReview[[4]] K8s APIs, before forwarding it to the internal, insecure
loopback port.

![03-runtime_structure_secure_metrics.png](assets/gep29-03-runtime_structure_secure_metrics.png)

> _Fig.3: Runtime structure with secure metrics_

Reviewer 2: Not sure if this is really relevant in the context of the GEP. Secure observability communication is something that affects several components. It does not seem to add any value to this GEP. In other words, I would prefer if we remove this section.

Author: I extracted the secure metrics section into external document and referenced it from the GEP. I removed the details from here.

## Discussion and Limitations
### Alternatives
The following existing 3rd party PVC autoscalers were evaluated:
- [topolvm/pvc-autoresizer](https://github.com/topolvm/pvc-autoresizer)
- [lorenzophys/pvc-autoscaler](https://github.com/lorenzophys/pvc-autoscaler)
- [DevOps-Nirvana/Kubernetes-Volume-Autoscaler](https://github.com/DevOps-Nirvana/Kubernetes-Volume-Autoscaler)

**Note:** The evaluation of each autoscaling option also accounts for the utility it adds to the envisioned future

> scenario of offering PVC autoscaling service to shoot (customer) volumes.

Reviewer 2: I find it unfortunate that you get to know more details about `pvc-autoscaler` by reading this section and between its lines compared to all previous sections. Would it make sense to discuss `pvc-autoscaler` and its features in more detail in a section of the GEP?

Author: I'd like to avoid duplicating `pvc-autoscaler`'s documentation in the GEP. If the `pvc-autoscaler` documentation is incomplete, I'd rather make the necessary changes there.

On the other hand, the GEP should indeed describe the specifics of how `pvc-autoscaler` will be applied in our concrete case. The thing is, because that's basic stuff, and very obvious to me, I'm quite blind to omissions on that part. So, please take a look a the new introductory paragraph I added to the GEP (available with next push), and let me know what I'm missing.

Reviewer 2: I agree that duplicating documentation is not great, but then again you spend 70 lines on why the alternatives are not good and we can only read between the lines why pvc-autoscaler is better.

You might have worked closer with pvc-autoscaler and hence you may already know this, but everybody else just sees you bashing a couple of projects without explicitly writing what you want/what pvc-autoscaler provides.

From my point of view, it would be a good idea to have a short section for pvc-autoscaler here as well listing what we think is relevant.

Author: I added a brief description of `pvc-autoscaler` in the start of the "Discussion and Limitations" section. Will be available with next push.

**Recommendation:**
It has a major disadvantage, compared to `pvc-autoscaler`.

#### Alternative: Business as usual
**Overview:**
The existing approach of statically allocating storage has the average observability volume severely underutilised, yet
bears a risk of high-load shoots exhausting their allocated storage. 

**Recommendation:**
The use of `pvc-autoscaler` would result in a substantial economy of storage space. 

**Details:**
The average observability volume utilisation, observed on a set of typical Gardener shoots, running with the current
default settings, was:
- Prometheus average utilisation: 16.6%; &plusmn;1&sigma;: (8.0 - 25.2)
- Vali average utilisation: 8.3%; &plusmn;1&sigma;: (4.2 - 12.3)

The above aggregations include both shoot and seed volumes.

### Limitations
#### Reliance on Prometheus
`pvc-autoscaler` relies on metrics from the seed's cache Prometheus instance. The availability of the otherwise simple,
and lightly and predictably loaded, autoscaler is thus limited to the availability level of the complex Prometheus, which
may be subject to load spikes and autoscaling evictions. By itself, this limitation is deemed acceptable, as the trigger

> thresholds, configured for PVC autoscaling, usually afford tolerance to substantial delays.

Reviewer 2: Maybe I did not see it, but does the GEP actually give some idea what the trigger threshold should be? Is it dependent on the infrastructure as volume expansion might have different performance characteristics depending on the infrastructure?

Author: Excellent point! I'm missing an introduction!

I added a short introductory paragraph early in the document, right under the "Proposal" title. Will be available with next push.

Author: To highlight the few discussion threads which remain open, I'm closing threads which appear to have been addressed. If you find that anything needs more discussion/work, please reopen.

Using the seed's cache Prometheus instance creates a dependency loop - `pvc-autoscaler` needs `prometheus-cache` in order
to perform scaling, and `prometheus-cache` needs `pvc-autoscaler` to scale its database volume. That dependency results
in one mode of persistent failure:
1. `prometheus-cache` exhausts its storage before `pvc-autoscaler` resizes it
2. `prometheus-cache` enters a fault loop, due to inability ot expand database
3. `pvc-autoscaler` enters fault loop, due to driving metrics signal unavailable
4. At this point, autoscaling seizes for all PVs on the seed

To mitigate the risk of such failure, a larger initial storage request is used for `prometheus-cache` - 6Gi. It offers a
1Gi headroom over the 5Gi storage retention limit, configured on `prometheus-cache`. Field data suggests 1Gi should
suffice to accommodate the occasional usage spikes which exceed the data retention limit.

#### Latency
The pipeline which propagates the volume metrics, driving `pvc-autoscaler`, is as follows:
1. `kubelet` collects PV state data and publishes resource metrics. Default poll interval: 1 minute

>    ([source](https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/))

Reviewer 2: The link is not very helpful. Maybe https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/ is better when pointing out that likely you are referring to `volumeStatsAggPeriod`.

Author: I suspect that the page was updated since I originally referenced it. I replaced the link with the API reference one you provided.

I wonder if, instead of linking to external documentation directly, I should be linking to copies in the Internet Archive.

2. `metrics-server` polls `kubelet` resource metrics. Default poll interval: 1 minute
   ([source](https://github.com/kubernetes-sigs/metrics-server/blob/master/FAQ.md))
3. `kube-apiserver` forwards metrics API calls to `metrics-server`, which is the registered APIService for metrics. Call
   forwarding is instantaneous, so the delay is minimal.
4. The seed's cache Prometheus polls the metrics provided by `kube-apiserver`. Default poll interval: 1 minute.
5. `pvc-autoscaler` polls the seed's cache Prometheus instance. Poll interval: 1 minute.

The reaction time, until `pvc-autoscaler` **initiates** a volume resize, can thus exceed 4 minutes. The resize operation
itself, including the file system resize, can further take a few minutes. The cumulative delay, roughly estimated to up
to 10 minutes, is mitigated via an appropriate safety margin, provided by the choice of utilisation level, configured

> as a scaling trigger threshold.

Reviewer 2: How is the trigger threshold chosen? How is it configured?

Author: I am now mentioning the threshold and its value (~90%~ correction: 10%) in the introductory paragraph. It is the `pvc-autoscaler`'s default. I reviewed it, and thought it reasonable. I did not come to it via some sort of optimisation calculation. I did however verify against Prometheus data, that it suffices to accommodate storage allocation spikes.

It can be configured via annotation on individual PVCs. However, I intend to leave it unspecified, and use the default.

Author: >it suffices to accommodate storage allocation spikes

As I read this, I realised that it's not quite true. I am aware of one category of spikes which can overwhelm the autoscaler with the proposed settings: Prometheus DB compaction causes limited, but particularly sharp surges in storage use (those are the spikes, subtly referred to in the _Limitations>Reliance on Prometheus_ section). For `cache-prometheus`, which is the critical case as it can result in persistent failure, I've mitigated those by setting a minimal volume size to 6Gi. Which results in scaling threshold of 10%=600Mi+ - enough to accommodate a spike.

I've been looking into an idea to enhance `pvc-autoscaler` with a behaviour which suits workloads with such sharp spikes, but I haven't reached a stage where I'd want to reflect it in the GEP.

Author: I'm in the process of adding support for an absolute `min-threshold` value (in bytes) to `pvc-autoscaler`. By setting an absolute minimum to the threshold, the aforementioned issue with sharp spikes when the volume is small, is completely resolved. I reflected this change by adding a new column to the "Initial volume size and MaxAllowed". For which, as an afterthought, a different title may now be in order.
